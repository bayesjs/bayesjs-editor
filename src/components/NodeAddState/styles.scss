.node-add-state {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 10px 0px;

  &__input {
    flex-grow: 1;
  }

  &__button {
    margin-left: 10px;
  }
}
