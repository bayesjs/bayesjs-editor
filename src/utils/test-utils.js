import {
  compose,
  filter,
  isEmpty,
  join,
  map,
  not,
} from 'ramda';

import { kebabCase } from 'change-case';
import removeAccents from 'remove-accents';

const formatWord = compose(removeAccents, kebabCase);
const notEmpty = compose(not, isEmpty);
const filterSubItems = filter(notEmpty);
const formatSubItems = compose(join('-'), map(formatWord));

export const getComponentTestId = (componentName, ...subItems) => { //eslint-disable-line
  const compName = formatWord(componentName);
  const filtredItems = filterSubItems(subItems);

  return notEmpty(filtredItems) ? `${compName}__${formatSubItems(filtredItems)}` : compName;
};
