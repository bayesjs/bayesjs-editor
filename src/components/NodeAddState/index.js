import React, { useState, useCallback } from 'react';
import Icon from 'components/Icon';
import Button from 'components/Button';
import PropTypes from 'prop-types';
import {
  pipe,
  path,
  complement,
  isEmpty,
  trim,
  both,
} from 'ramda';
import { isEnterKey } from 'utils/event';
import { bem } from 'utils/styles';
import styles from './styles.scss';

const componentClassName = bem(styles);
const isNotEmpty = complement(isEmpty);
const pathTargetValue = path(['target', 'value']);

const NodeAddState = ({ onAddState }) => {
  const [state, setState] = useState('');
  const onChange = useCallback(pipe(pathTargetValue, setState), [setState]);
  const onValidAndSave = useCallback(both(isNotEmpty, onAddState), [onAddState]);
  const onSave = useCallback(() => onValidAndSave(trim(state)) && setState(''), [state, onAddState]);
  const onKeyUp = useCallback(both(isEnterKey, onSave), [state, onAddState]);

  return (
    <div className={componentClassName}>
      <input
        className={componentClassName.element('input').toString()}
        type="text"
        placeholder="Novo estado"
        value={state}
        onChange={onChange}
        onKeyUp={onKeyUp}
      />
      <Button
        className={componentClassName.element('button').toString()}
        onClick={onSave}
        title="Adicionar Estado"
      >
        <Icon name="plus" />
      </Button>
    </div>
  );
};

NodeAddState.propTypes = {
  onAddState: PropTypes.func.isRequired,
};

export default NodeAddState;
