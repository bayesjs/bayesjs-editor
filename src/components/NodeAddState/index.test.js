import * as eventUtils from '@utils/event';

import { fireEvent, render } from '@testing-library/react';

import React from 'react';
import { getComponentTestId } from '@utils/test-utils';
import NodeAddState from './index';

const renderComponent = props =>
  render(
    <NodeAddState {...props} />,
  );

describe('NodeAddState Component', () => {
  const defaultProps = {
    onAddState: jest.fn(() => true),
  };
  let component;

  beforeEach(() => {
    defaultProps.onAddState.mockClear();
    component = renderComponent(defaultProps);
  });

  it('matches snapshot', () => {
    expect(component.container.firstChild).toMatchSnapshot();
  });

  describe('When keyup in input', () => {
    const { isEnterKey } = eventUtils;

    describe('and is "enter" key', () => {
      beforeEach(() => {
        const { getByPlaceholderText } = component;
        const input = getByPlaceholderText('Novo estado');

        eventUtils.isEnterKey = () => true;
        fireEvent.change(input, { target: { value: 'new state' } });
        fireEvent.keyUp(input);
      });

      afterAll(() => {
        eventUtils.isEnterKey = isEnterKey;
      });

      it('calls onAddState function', () => {
        expect(defaultProps.onAddState).toBeCalled();
      });
    });

    describe('and is not "enter" key', () => {
      beforeEach(() => {
        const { getByPlaceholderText } = component;
        const input = getByPlaceholderText('Novo estado');

        eventUtils.isEnterKey = () => false;
        fireEvent.change(input, { target: { value: 'new state' } });
        fireEvent.keyUp(input);
      });

      afterAll(() => {
        eventUtils.isEnterKey = isEnterKey;
      });

      it('does not call onAddState function', () => {
        expect(defaultProps.onAddState).not.toBeCalled();
      });
    });
  });

  describe('When clicked in add state button', () => {
    describe('and has a valid state', () => {
      const newState = 'new state';

      beforeEach(() => {
        const { getByPlaceholderText, getByTestId } = component;
        const input = getByPlaceholderText('Novo estado');
        const addButton = getByTestId(getComponentTestId('Button', 'Adicionar Estado'));

        fireEvent.change(input, { target: { value: newState } });
        fireEvent.click(addButton);
      });

      it('calls onAddState function', () => {
        expect(defaultProps.onAddState).toBeCalledWith(newState);
      });
    });

    describe('and has a invalid state', () => {
      const newState = '';

      beforeEach(() => {
        const { getByPlaceholderText, getByTestId } = component;
        const input = getByPlaceholderText('Novo estado');
        const addButton = getByTestId(getComponentTestId('Button', 'Adicionar Estado'));

        fireEvent.change(input, { target: { value: newState } });
        fireEvent.click(addButton);
      });

      it('calls onAddState function', () => {
        expect(defaultProps.onAddState).not.toBeCalled();
      });
    });
  });
});
