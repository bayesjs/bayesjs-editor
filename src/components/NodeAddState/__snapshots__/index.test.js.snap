// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeAddState Component matches snapshot 1`] = `
<div
  class=""
>
  <input
    class="__input"
    placeholder="Novo estado"
    type="text"
    value=""
  />
  <button
    class="__button"
    data-testid="button__adicionar-estado"
    title="Adicionar Estado"
    type="button"
  >
    <i
      class="fa fa-plus"
    />
  </button>
</div>
`;
